packages/core/src/api/clipboard/fromClipboard/pasteExtension.ts

```diff
@@ -8,9 +8,101 @@ import {
   StyleSchema,
 } from "../../../schema/index.js";
 import { nestedListsToBlockNoteStructure } from "../../parsers/html/util/nestedLists.js";
+import { markdownToHTML } from "../../parsers/markdown/parseMarkdown.js";
 import { acceptedMIMETypes } from "./acceptedMIMETypes.js";
 import { handleFileInsertion } from "./handleFileInsertion.js";
 import { handleVSCodePaste } from "./handleVSCodePaste.js";
+import { isMarkdown } from "../../parsers/markdown/detectMarkdown.js";
+import { EditorView } from "prosemirror-view";
+
+function convertHtmlToBlockNoteHtml(html: string) {
+  const htmlNode = nestedListsToBlockNoteStructure(html.trim());
+  return htmlNode.innerHTML;
+}
+
+function convertMarkdownToBlockNoteHtml(markdown: string) {
+  return markdownToHTML(markdown).then((html) => {
+    return convertHtmlToBlockNoteHtml(html);
+  });
+}
+
+function defaultPasteHandler({
+  view,
+  event,
+  editor,
+  pasteBehavior = "prefer-markdown",
+}: {
+  view: EditorView;
+  event: ClipboardEvent;
+  editor: BlockNoteEditor<any, any, any>;
+  pasteBehavior?: "prefer-markdown" | "prefer-html";
+}) {
+  let format: (typeof acceptedMIMETypes)[number] | undefined;
+  for (const mimeType of acceptedMIMETypes) {
+    if (event.clipboardData!.types.includes(mimeType)) {
+      format = mimeType;
+      break;
+    }
+  }
+
+  if (!format) {
+    return true;
+  }
+
+  if (format === "vscode-editor-data") {
+    handleVSCodePaste(event, view);
+    return true;
+  }
+
+  if (format === "Files") {
+    handleFileInsertion(event, editor);
+    return true;
+  }
+
+  const data = event.clipboardData!.getData(format);
+
+  if (format === "blocknote/html") {
+    // Is blocknote/html, so no need to convert it
+    view.pasteHTML(data);
+    return true;
+  }
+
+  if (format === "text/markdown") {
+    convertMarkdownToBlockNoteHtml(data).then((html) => {
+      view.pasteHTML(html);
+    });
+    return true;
+  }
+
+  if (format === "text/html") {
+    if (pasteBehavior === "prefer-markdown") {
```

Comment:
code style feedback:

instead of having these checks spread here and below (L95), I feel like it would be cleaner to check the flag earlier (or even in a separate "middleware" function) that would just set the format to "text/markdown" and then let the default behavior paste that. makes sense / wdyt? (I didn't completely think it out)